rogue/src/player.rs

```diff
@@ -273,86 +275,234 @@ pub fn player_input(gs: &mut State, ctx: &mut Rltk) -> RunState {
+        let mut searchpt = Point{x: pt.x + dx, y: pt.y + dy};
+        // Search along the ray from the player position in the direction of
+        // movement for any meele target. We stop the search as soon as we find:
+        //   - A tile outside the player's view.
+        //   - Any blocked tile that is not meele attackable.
+        //   - A meele attack target.
+        // In the final case, we *know* the prevous search tile is unblocked,
+        // because we encountered it in the seatch.
+        loop {
+            let search_idx = map.xy_idx(searchpt.x, searchpt.y);
+            let tile_not_visible =  !vs.visible_tiles.contains(&searchpt);
+            let is_blocked = map.blocked[search_idx];
+            // Again, don't count invisible meele targets in the dash tile.
+            let dash_targets = get_meele_targets_in_tile(
+                &*map, search_idx, &combat_stats, &invisibles, false
+            );
+            if tile_not_visible {
+                break;
+            } else if dash_targets.is_empty() && is_blocked {
+                break
+            // Found a target: DASH!
+            } else if !dash_targets.is_empty() {
+                special.expend();
+                meele_buffer.request_many(*player, &dash_targets, true);
+                moves.insert(
+                    *player,
+                    WantsToMoveToPosition {
+                        pt: Point{x: searchpt.x - dx, y: searchpt.y - dy},
+                        force: false
+                    }
+                ).expect("Failed to insert dash meele attack move.");
+                animation_buffer.request(AnimationRequest::AlongRay {
+                    source_x: pt.x,
+                    source_y: pt.y,
+                    target_x: searchpt.x - dx,
+                     target_y: searchpt.y - dy,
+                     fg: RGB::named(rltk::PURPLE),
+                     bg: playerrender.bg,
+                     glyph: playerrender.glyph,
+                     until_blocked: true
+                });
+                return RunState::PlayerTurn
+            }
+            searchpt = Point {x: searchpt.x + dx, y: searchpt.y + dy};
         }
-        // IMPORTANT: Keeps the players Position component synchronized with
-        // their position as a resource in the ECS.
-        ppos.x = pos.x;
-        ppos.y = pos.y;
     }
-    return RunState::PlayerTurn;
+
+    // If the destination tile is unblocked, we can move the player into it, and
+    // then pass the turn.
+    if !destination_is_blocked {
+        moves.insert(
+            *player,
+            WantsToMoveToPosition {pt: Point{x: pt.x + dx, y: pt.y + dy}, force: false}
+        ).expect("Failed to insert player meele attack move.");
+        return RunState::PlayerTurn;
+    }
+
+    // If the desination tile is blocked, yet no meele happened, then you're
+    // trying to move against a wall or something and we give you a freebie,
+    // don't pass the turn.
```

Comment:
Oooh this feels like fun insider knowledge - using the dash as a kind of free-movement action.